tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+                        let overflow_flag_lut = self.key.generate_lookup_table(|block| {
+                            let input_carry = block & 1;
+                            if input_carry == 1 {
+                                (block >> 3) & 1
+                            } else {
+                                (block >> 2) & 1
+                            }
+                        });
+
+                        self.key
+                            .apply_lookup_table_assign(overflow_flag_block, &overflow_flag_lut);
+                    });
+                }
+            });
+
+            return match requested_flag {
+                OutputFlag::None => None,
+                OutputFlag::Overflow => {
+                    assert!(
+                        overflow_flag.is_some(),
+                        "internal error, overflow_flag should exist"
+                    );
+                    overflow_flag.map(BooleanBlock::new_unchecked)
+                }
+                OutputFlag::Carry => {
+                    carry.degree = Degree::new(1);
+                    Some(BooleanBlock::new_unchecked(carry))
+                }
+            };
+        }
+
+        // 1_X parameters
+        //
+        // Same idea as other algorithms, however since we have 1 bit per block
+        // we do not have to resolve any inner propagation but it adds one more
+        // sequential PBS when we are interested in the OverflowFlag
+        if self.key.message_modulus.0 == 2 {
+            fn block_add_assign_returning_carry(
+                sks: &ServerKey,
+                lhs: &mut Ciphertext,
+                rhs: &Ciphertext,
+                carry: &Ciphertext,
+            ) -> Ciphertext {
+                sks.key.unchecked_add_assign(lhs, rhs);
+                sks.key.unchecked_add_assign(lhs, carry);
+                let (carry, message) = rayon::join(
+                    || sks.key.carry_extract(lhs),
+                    || sks.key.message_extract(lhs),
+                );
+
+                *lhs = message;
+
+                carry
+            }
+            let num_blocks = lhs.len();
+            for (lhs_b, rhs_b) in lhs[..num_blocks - 1]
+                .iter_mut()
+                .zip(rhs[..num_blocks - 1].iter())
+            {
+                carry = block_add_assign_returning_carry(self, lhs_b, rhs_b, &carry);
+            }
+
+            let mut output_carry = block_add_assign_returning_carry(
+                self,
+                &mut lhs[num_blocks - 1],
+                &rhs[num_blocks - 1],
+                &carry,
+            );
+
+            return match requested_flag {
+                OutputFlag::None => None,
+                OutputFlag::Overflow => {
+                    let overflowed = self.key.not_equal(&output_carry, &carry);
+                    Some(BooleanBlock::new_unchecked(overflowed))
+                }
```

Comment:
I'm having a hard time figuring out how this works for signed values ?

Reply:
You mean why checking the difference between input and output carry of the last bit works to check for overflow ?

Reply:
Re reading the code I might have mistaken with something else, but sure a comment here could be welcome as it's not immediately clear how it works, the algo is still complex